src/EditorFeatures/Test/Squiggles/AbstractSquiggleProducerTests.cs

```diff
@@ -23,43 +25,20 @@ public static class SquiggleUtilities
     {
         internal static List<ITagSpan<IErrorTag>> GetErrorSpans(
             TestWorkspace workspace,
-            ImmutableDictionary<string, ImmutableArray<DiagnosticAnalyzer>> analyzerMap = null)
+            Dictionary<string, DiagnosticAnalyzer[]> analyzerMap = null)
         {
-            var registrationService = workspace.Services.GetService<ISolutionCrawlerRegistrationService>();
-            registrationService.Register(workspace);
-
-            var listener = new AsynchronousOperationListener();
-            var listeners = AsynchronousOperationListener.CreateListeners(
-                ValueTuple.Create(FeatureAttribute.DiagnosticService, listener),
-                ValueTuple.Create(FeatureAttribute.ErrorSquiggles, listener));
-
-            var optionsService = workspace.Services.GetService<IOptionService>();
-
-            var analyzerService = analyzerMap == null || analyzerMap.Count == 0
-                ? new TestDiagnosticAnalyzerService(DiagnosticExtensions.GetCompilerDiagnosticAnalyzersMap())
-                : new TestDiagnosticAnalyzerService(analyzerMap);
-
-            var diagnosticService = new DiagnosticService(SpecializedCollections.SingletonEnumerable<IDiagnosticUpdateSource>(analyzerService), listeners);
-
-            var document = workspace.Documents.First();
-            var buffer = document.GetTextBuffer();
-
-            var foregroundService = workspace.GetService<IForegroundNotificationService>();
-            var taggerProvider = new DiagnosticsSquiggleTaggerProvider(optionsService, diagnosticService, foregroundService, listeners);
-            var tagger = taggerProvider.CreateTagger<IErrorTag>(buffer);
-            using (var disposable = tagger as IDisposable)
+            using (var wrapper = new DiagnosticTaggerWrapper(workspace, analyzerMap))
             {
-                var service = workspace.Services.GetService<ISolutionCrawlerRegistrationService>() as SolutionCrawlerRegistrationService;
-                service.WaitUntilCompletion_ForTestingPurposesOnly(workspace, ImmutableArray.Create(analyzerService.CreateIncrementalAnalyzer(workspace)));
+                var tagger = wrapper.TaggerProvider.CreateTagger<IErrorTag>(workspace.Documents.First().GetTextBuffer());
+                using (var disposable = tagger as IDisposable)
+                {
+                    wrapper.WaitForTags();
 
-                listener.CreateWaitTask().PumpingWait();
+                    var snapshot = workspace.Documents.First().GetTextBuffer().CurrentSnapshot;
```

Comment:
This is test code.  And all these tests are written as if they operate on a single document.